proposals/p2022.md

```diff
@@ -0,0 +1,276 @@
+but uses are invalid, including when they cause ambiguous name lookup errors. If
+attempted to be used, a compiler error will be shown to the user, instructing
+them to either remove the `unused` qualifier or remove the use.
+
+The inverse, where a name is not qualified by the `unused` qualifier, but never
+used, will cause the compiler to emit a warning diagnostic, informing the user
+that a given name was not used and suggesting to either remove the binding or
+mark it as unused.
+
+### Examples
+
+Discarded Name : Function Parameter
+
+```carbon
+// Interface
+fn Sum(x: List(Int), size: Int) -> Int;
+
+// Implementation that doesn't use the size parameter
+fn Sum(x: List(Int), _: Int) -> Int;
+```
+
+Discarded Name : Case Binding
+
+```carbon
+fn Bar() -> (Int, Int);
+fn Foo() -> Int {
+  match (Bar()) {
+    case (42, y: Int) => {
+      y
+    }
+    case (x: Int, _: Int => {
+      x
+    }
+  }
+}
+```
+
+Discarded Name : Variable Binding
+
+```carbon
+fn Bar() -> (Int, Int);
+fn Foo() -> Int {
+  var (x: Int, _: Int) = Bar();
+  return x;
+}
+```
+
+Unused Name : Function Parameter
+
+```carbon
+// Interface
+fn Sum(x: List(Int), size: Int) -> Int;
+
+// Implementation that doesn't use the size parameter
+fn Sum(x: List(Int), unused size: Int) -> Int;
+```
+
+Unused Name : Case Binding
+
+```carbon
+fn Bar() -> (Int, Int);
+fn Foo() -> Int {
+  match (Bar()) {
+    case (42, y: Int) => {
+      return y;
+    }
+    case (x: Int, unused y: Int) => {
+      return x;
+    }
+  }
+}
+```
+
+Unused Name : Variable Binding
+
+```carbon
+fn Bar() -> (Int, Int);
+fn Foo() -> Int {
+  var (x: Int, unused y: Int) = Bar();
+  return x;
+}
+```
+
+## Rationale
+
+-   [Code that is easy to read, understand, and write](/docs/project/goals.md#code-that-is-easy-to-read-understand-and-write)
+-   Carbon should not use symbols that are difficult to type, see, or
+    differentiate from similar symbols in commonly used contexts.
+
+-   Syntax should be easily parsed and scanned by any human in any development
+    environment, not just a machine or a human aided by semantic hints from an
+    IDE.
+-   Explicitness must be balanced against conciseness, as verbosity and ceremony
+    add cognitive overhead for the reader, while explicitness reduces the amount
+    of outside context the reader must have or assume.
+
+This proposal uses the underscore, `_`, character to denote an unused value, a
+meaning used across various other programming languages. A lone underscore
+character only has a single meaning in Carbon and will be unambiguous across
+contexts.
+
+Both syntaxes are easily read by humans, either by seeing the `_` chacter alone,
+or by introducing a keyword that allows code to read naturally such as
+`unused size : Int`.
+
+The inclusion of two syntaxes allows authors to decide when they will favor
+conciseness or explicitness over the other.
+
+## Alternatives considered
```

Comment:
Another rule that was considered was saying any parameter named starting with an `_` is unused. See https://github.com/carbon-language/carbon-lang/issues/476#issuecomment-826068624 . Please include the alternative and the rationale why it was not chosen (both what was stated in the issue and in https://github.com/carbon-language/carbon-lang/pull/2022/files#r945934384 ).

Reply:
Added an alternative for this -- reading over #476 it seemed the main argument against is that its "clever". Let me know if I should add something more

Reply:
Should also include something about https://github.com/carbon-language/carbon-lang/pull/2022/files#r945934384 : "reserving" the name for later use (and avoiding accidents where there happens to be a matching name for lookup to find).

Reply:
I didn't include anything along those lines because if names prefixed with `_` are discarded, there's nothing for them to ever conflict with or accidentally resolve to. Do you mean to include a pro that it can't lead to accidents?

Reply:
It is a possible advantage of the `unused` keyword approach that you can reserve a name, and a potential disadvantage of the `_`-prefix approach that you can't.

Reply:
Gotcha -- thanks for the clarification. I'll update it this evening.

Reply:
@chandlerc Maybe you could explain what was bad about being "clever" in https://github.com/carbon-language/carbon-lang/issues/476#issuecomment-826068624 ?

Reply:
Good call, suggested more specific wording below.